File: sdks/kotlin/src/test/kotlin/io/github/mikemackintosh/crowdcontrol/ValidateTest.kt
package io.github.mikemackintosh.crowdcontrol

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ValidateTest {

    @Test
    fun flagsUnknownField() {
        val p = parse("""forbid "r" { user.unknown == "x" }""")
        val schema = Schema(mapOf("user.name" to FieldType.STRING))
        val warnings = validatePolicies(listOf(p), schema)
        assertEquals(1, warnings.size)
        assertTrue(warnings[0].message.contains("not found"))
    }

    @Test
    fun allowsKnownField() {
        val p = parse("""forbid "r" { user.name == "alex" }""")
        val schema = Schema(mapOf("user.name" to FieldType.STRING))
        val warnings = validatePolicies(listOf(p), schema)
        assertTrue(warnings.isEmpty())
    }

    @Test
    fun flagsNumericOpOnString() {
        val p = parse("""forbid "r" { user.name < 5 }""")
        val schema = Schema(mapOf("user.name" to FieldType.STRING))
        val warnings = validatePolicies(listOf(p), schema)
        assertEquals(1, warnings.size)
        assertTrue(warnings[0].message.contains("<"))
    }

    @Test
    fun flagsQuantifierOnNonList() {
        val p = parse("""forbid "r" { any user.name == "x" }""")
        val schema = Schema(mapOf("user.name" to FieldType.STRING))
        val warnings = validatePolicies(listOf(p), schema)
        assertTrue(warnings.any { it.message.contains("expected list") })
    }

    @Test
    fun flagsCountOnWrongType() {
        val p = parse("""forbid "r" { count(user.name) > 1 }""")
        val schema = Schema(mapOf("user.name" to FieldType.STRING))
        val warnings = validatePolicies(listOf(p), schema)
        assertTrue(warnings.any { it.message.contains("expected list or number") })
    }

    @Test
    fun flagsUnknownInterpolation() {
        val p = parse(
            """
            forbid "r" {
                user.role == "intern"
                message "bad: {user.nope}"
            }
            """.trimIndent()
        )
        val schema = Schema(mapOf("user.role" to FieldType.STRING))
        val warnings = validatePolicies(listOf(p), schema)
        assertTrue(warnings.any { it.field == "user.nope" })
    }

    @Test
    fun mapTypeAllowsAnySubpath() {
        val p = parse("""forbid "r" { user.profile.country == "US" }""")
        val schema = Schema(mapOf("user.profile" to FieldType.MAP))
        val warnings = validatePolicies(listOf(p), schema)
        assertTrue(warnings.isEmpty())
    }

    @Test
    fun formatWarningsRendersEachOnALine() {
        val warnings = listOf(
            SchemaWarning(rule = "r1", field = "x", message = "bad x"),
            SchemaWarning(rule = "r2", field = "y", message = "bad y"),
        )
        val text = formatWarnings(warnings)
        assertTrue(text.contains("r1: bad x"))
        assertTrue(text.contains("r2: bad y"))
    }
}

File: sdks/php/tests/validate_test.php
<?php

declare(strict_types=1);

use MikeMackintosh\CrowdControl\CrowdControl;
use MikeMackintosh\CrowdControl\Schema;
use MikeMackintosh\CrowdControl\Validate;

function test_validate_unknown_field(): void
{
    $policies = [CrowdControl::parse('forbid "r" { unknown.field == "x" }')];
    $schema = new Schema(['resource.type' => 'string']);
    $warnings = Validate::validatePolicies($policies, $schema);
    $found = false;
    foreach ($warnings as $w) {
        if ($w->field === 'unknown.field') {
            $found = true;
        }
    }
    assert_true($found, 'expected warning for unknown.field');
}

function test_validate_known_field_no_warning(): void
{
    $policies = [CrowdControl::parse('forbid "r" { resource.type == "x" }')];
    $schema = new Schema(['resource.type' => 'string']);
    $warnings = Validate::validatePolicies($policies, $schema);
    assert_count(0, $warnings);
}

function test_validate_operator_mismatch(): void
{
    $policies = [CrowdControl::parse('forbid "r" { user.name < 5 }')];
    $schema = new Schema(['user.name' => 'string']);
    $warnings = Validate::validatePolicies($policies, $schema);
    $found = false;
    foreach ($warnings as $w) {
        if (str_contains($w->message, '<')) {
            $found = true;
        }
    }
    assert_true($found);
}

function test_validate_count_on_string(): void
{
    $policies = [CrowdControl::parse('forbid "r" { count(x) > 3 }')];
    $schema = new Schema(['x' => 'string']);
    $warnings = Validate::validatePolicies($policies, $schema);
    $found = false;
    foreach ($warnings as $w) {
        if (str_contains($w->message, 'count')) {
            $found = true;
        }
    }
    assert_true($found);
}

function test_validate_interpolation_unknown_field(): void
{
    $src = 'forbid "r" { y == "x" message "hi {user.name}" }';
    $policies = [CrowdControl::parse($src)];
    $schema = new Schema(['y' => 'string']);
    $warnings = Validate::validatePolicies($policies, $schema);
    $found = false;
    foreach ($warnings as $w) {
        if ($w->field === 'user.name') {
            $found = true;
        }
    }
    assert_true($found);
}
